### PR TITLE
apt: explicitly query apt-cache for a single package

### PR DESCRIPTION
**Pull request by OddBloke**
## Description

`apt-cache` treats arguments which don't correspond to package names as
patterns: this means that Chef can end up attempting to install a
different set of packages to the ones requested if the request includes
an absent package with pattern-matching characters (e.g. `lua5.3` in
Ubuntu if universe is disabled will match `liblua5.3-dev` and others).

This commit modifies the `apt-cache` calls in the apt package provider
which pass a package name to wrap that package name in '^...$', ensuring
that `apt-cache` will attempt to select a single package regardless of
whether the package name contains pattern-matching characters.

An alternative would be to set `APT::Cmd::Pattern-Only` as `true` when
executing these commands: not all supported Ubuntu/Debian versions have
an apt which supports that configuration option, so this commit's
approach is more universal.

## Related Issue

Fixes: #12944

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [ ] I have run the pre-merge tests locally and they pass.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by tpowell-progress** [comment]
Concerns:
1) Are any users relying on the implicit pattern match explicitly in their workflows and will be broken by this change?
2) Are any users implicitly using this pattern and will break because their package spec is inadvertently an inexact match?
3) How does this behavior compare with other *nix package managers? (see: chef/provider/package)

For 1 & 2 we need to assess risk, but likely a breaking change that would wait until next major release?

**Comment by marcparadise** [comment]
@OddBloke thanks for taking the time to work on this PR.  We can't accept this change as-is, because of the risk of breaking existing usage outlined above.  

Would you willing to update this to use a property on the apt-package resource to control whether pattern matching is enabled, defaulting to `true` for compatibility?

**Comment by jaymzh** [comment]
So you can add a property the way it's done here, in the resource:

https://github.com/chef/chef/blob/main/lib/chef/resource/apt_package.rb#L63

So you might make one called `anchor_package_name_patterns` for example, in that file.

Which just makes a variable you can use, `new_resource.anchor_package_name_patterns`, which is used like this:

https://github.com/chef/chef/blob/main/lib/chef/provider/package/apt.rb#L169

basically, you'd add the property and then adapt your existing code to use an if:

```ruby
if new_resource.anchor_package_name_patterns
   # your new code here
else
  # your old code here
end
```

Then peopel will just do:

```ruby
package 'whatever' do
  anchor_package_name_patterns true
  action :upgrade
end
```

**Comment by OddBloke** [comment]
Thanks for the pointers! This is still on my list, but I'm having a hectic time at the moment so it'll still be a little while before I can push this forward.

**Comment by jaymzh** [comment]
ping, @OddBloke

**Comment by tpowell-progress** [comment]
This will be superseded by @neha-p6's PR, so closing
